File: sdk/flutter/praxy_sdk_gen/lib/src/spec.dart
/// Reading `docs/openapi/v1.json` — just enough of OpenAPI to describe Praxy's own
/// document, deliberately not a general-purpose parser.
library;

/// One operation, reduced to what a Dart method needs.
final class Operation {
  Operation({
    required this.operationId,
    required this.method,
    required this.path,
    required this.pathParameters,
    required this.queryParameters,
    required this.bodyProperties,
    required this.requiredBodyProperties,
    required this.responseSchema,
    required this.summary,
  });

  final String operationId;
  final String method;
  final String path;
  final List<String> pathParameters;
  final List<Parameter> queryParameters;

  /// Request-body properties, flattened into named method parameters rather than a
  /// request object. A caller writes `create(email: 'a@b.com')`, not
  /// `create(ServerCreateUserRequest(email: 'a@b.com'))` — matching every
  /// hand-written service in `praxy_core`, so generated and hand-written methods
  /// are indistinguishable at the call site.
  final Map<String, SchemaRef> bodyProperties;
  final Set<String> requiredBodyProperties;

  /// The success response's schema, or null for a 204.
  final SchemaRef? responseSchema;
  final String? summary;

  /// The method name: `users.updateEmail` → `updateEmail`.
  String get methodName =>
      operationId.contains('.') ? operationId.split('.').last : operationId;
}

final class Parameter {
  Parameter({required this.name, required this.schema, required this.required});
  final String name;
  final SchemaRef schema;
  final bool required;
}

/// A resolved schema: either a reference to a named component or an inline type.
final class SchemaRef {
  SchemaRef({this.componentName, required this.type, this.itemType, this.format});

  /// Set when this is a `$ref` to `#/components/schemas/<name>`.
  final String? componentName;
  final String type;
  final SchemaRef? itemType;
  final String? format;
}

/// Parses the operations of one `operationId` namespace (`users.*`) out of the document.
List<Operation> parseNamespace(Map<String, dynamic> document, String namespace) {
  final paths = document['paths'] as Map<String, dynamic>;
  final operations = <Operation>[];

  for (final entry in paths.entries) {
    final item = entry.value as Map<String, dynamic>;
    for (final verb in const ['get', 'post', 'put', 'patch', 'delete']) {
      final op = item[verb];
      if (op is! Map<String, dynamic>) continue;
      final operationId = op['operationId'] as String?;
      if (operationId == null || !operationId.startsWith('$namespace.')) continue;

      final parameters = (op['parameters'] as List?) ?? const [];
      final pathParameters = <String>[];
      final queryParameters = <Parameter>[];
      for (final raw in parameters) {
        final p = raw as Map<String, dynamic>;
        final name = p['name'] as String;
        if (p['in'] == 'path') {
          pathParameters.add(name);
        } else if (p['in'] == 'query') {
          queryParameters.add(
            Parameter(
              name: name,
              schema: resolveSchema(p['schema'] as Map<String, dynamic>),
              required: p['required'] == true,
            ),
          );
        }
      }

      final bodySchema = _bodySchema(document, op);
      operations.add(
        Operation(
          operationId: operationId,
          method: verb.toUpperCase(),
          path: entry.key,
          pathParameters: pathParameters,
          queryParameters: queryParameters,
          bodyProperties: bodySchema.properties,
          requiredBodyProperties: bodySchema.required,
          responseSchema: _responseSchema(op),
          summary: op['summary'] as String?,
        ),
      );
    }
  }

  operations.sort((a, b) => a.operationId.compareTo(b.operationId));
  return operations;
}

final class _Body {
  _Body(this.properties, this.required);
  final Map<String, SchemaRef> properties;
  final Set<String> required;
}

_Body _bodySchema(Map<String, dynamic> document, Map<String, dynamic> op) {
  final content = (op['requestBody'] as Map<String, dynamic>?)?['content'] as Map<String, dynamic>?;
  final json = content?['application/json'] as Map<String, dynamic>?;
  final schema = json?['schema'] as Map<String, dynamic>?;
  if (schema == null) return _Body(const {}, const {});

  final resolved = _dereference(document, schema);
  final properties = (resolved['properties'] as Map<String, dynamic>?) ?? const {};
  return _Body(
    {
      for (final e in properties.entries)
        e.key: resolveSchema(e.value as Map<String, dynamic>),
    },
    ((resolved['required'] as List?) ?? const []).cast<String>().toSet(),
  );
}

SchemaRef? _responseSchema(Map<String, dynamic> op) {
  final responses = (op['responses'] as Map<String, dynamic>?) ?? const {};
  for (final status in const ['200', '201']) {
    final response = responses[status] as Map<String, dynamic>?;
    final schema = ((response?['content'] as Map<String, dynamic>?)?['application/json']
            as Map<String, dynamic>?)?['schema']
        as Map<String, dynamic>?;
    if (schema != null) return resolveSchema(schema);
  }
  return null;
}

Map<String, dynamic> _dereference(Map<String, dynamic> document, Map<String, dynamic> schema) {
  final ref = schema[r'$ref'] as String?;
  if (ref == null) return schema;
  final name = ref.split('/').last;
  final schemas = (document['components'] as Map<String, dynamic>)['schemas'] as Map<String, dynamic>;
  return schemas[name] as Map<String, dynamic>;
}

/// Turns one schema node into a [SchemaRef].
///
/// The `type` field is where .NET's document needs care: an integer property is
/// emitted as `type: ["integer", "string"]` with an integer `format`, even though the
/// wire only ever carries a number. Taking that union literally would give every
/// `total` in the SDK a `Object` type. `openapi-typescript` resolves the same union to
/// `number` for the console, so this resolves it to `int` — the same answer, reached
/// the same way, rather than a second opinion about the same document.
SchemaRef resolveSchema(Map<String, dynamic> schema) {
  final ref = schema[r'$ref'] as String?;
  if (ref != null) {
    return SchemaRef(componentName: ref.split('/').last, type: 'object');
  }

  final rawType = schema['type'];
  final types = switch (rawType) {
    String s => [s],
    List l => l.cast<String>(),
    _ => <String>['object'],
  };
  final type = types.firstWhere((t) => t != 'null' && t != 'string', orElse: () => types.first);

  return SchemaRef(
    type: type,
    format: schema['format'] as String?,
    itemType: type == 'array' && schema['items'] is Map<String, dynamic>
        ? resolveSchema(schema['items'] as Map<String, dynamic>)
        : null,
  );
}

File: sdk/generator/src/spec.mjs
// @ts-check
/**
 * Reads `docs/openapi/v1.json` into a small, language-neutral intermediate representation.
 *
 * This half of the generator is deliberately the only half that knows about OpenAPI, and it knows
 * nothing about any target language. Everything language-specific lives in `src/targets/`. That is
 * the whole reason this is one generator rather than one per language: adding a target is a new
 * file in `targets/`, not a new program that has to re-learn this document's quirks.
 */

/**
 * @typedef {object} SchemaRef
 * @property {string|null} componentName Set when this is a `$ref` to `#/components/schemas/<name>`.
 * @property {string} type
 * @property {SchemaRef|null} itemType
 * @property {string|null} format
 */

/**
 * @typedef {object} Parameter
 * @property {string} name
 * @property {SchemaRef} schema
 * @property {boolean} required
 * @property {string|null} description
 */

/**
 * @typedef {object} Operation
 * @property {string} operationId
 * @property {string} method
 * @property {string} path
 * @property {string[]} pathParameters
 * @property {Parameter[]} queryParameters
 * @property {Record<string, SchemaRef>} bodyProperties
 * @property {Set<string>} requiredBodyProperties
 * @property {SchemaRef|null} responseSchema Null for a 204.
 * @property {string|null} summary
 */

const HTTP_METHODS = ["get", "post", "put", "patch", "delete"];

/**
 * The operations in one `operationId` namespace (`users.*`), sorted so output is stable regardless
 * of the order paths happen to appear in the document.
 *
 * @param {any} document
 * @param {string} namespace
 * @returns {Operation[]}
 */
export function parseNamespace(document, namespace) {
  /** @type {Operation[]} */
  const operations = [];

  for (const [path, item] of Object.entries(document.paths)) {
    for (const method of HTTP_METHODS) {
      const op = /** @type {any} */ (item)[method];
      if (!op?.operationId?.startsWith(`${namespace}.`)) continue;

      /** @type {string[]} */
      const pathParameters = [];
      /** @type {Parameter[]} */
      const queryParameters = [];
      for (const parameter of op.parameters ?? []) {
        if (parameter.in === "path") pathParameters.push(parameter.name);
        else if (parameter.in === "query") {
          queryParameters.push({
            name: parameter.name,
            schema: resolveSchema(parameter.schema),
            required: parameter.required === true,
            description: parameter.description ?? null,
          });
        }
      }

      const body = bodySchema(document, op);
      operations.push({
        operationId: op.operationId,
        method: method.toUpperCase(),
        path,
        pathParameters,
        queryParameters,
        bodyProperties: body.properties,
        requiredBodyProperties: body.required,
        responseSchema: responseSchema(op),
        summary: op.summary ?? null,
      });
    }
  }

  operations.sort((a, b) => (a.operationId < b.operationId ? -1 : a.operationId > b.operationId ? 1 : 0));
  return operations;
}

/** The method name a target should use: `users.updateEmail` → `updateEmail`. */
/** @param {Operation} operation */
export function methodName(operation) {
  const dot = operation.operationId.lastIndexOf(".");
  return dot < 0 ? operation.operationId : operation.operationId.slice(dot + 1);
}

/**
 * Request-body properties, flattened rather than modelled as a request object, so a caller writes
 * `create(email: …)` instead of `create(CreateUserRequest(email: …))`. Every hand-written service
 * in every Praxy SDK reads that way, and generated methods should be indistinguishable from them.
 *
 * @param {any} document
 * @param {any} op
 */
function bodySchema(document, op) {
  const schema = op.requestBody?.content?.["application/json"]?.schema;
  if (!schema) return { properties: {}, required: new Set() };

  const resolved = dereference(document, schema);
  /** @type {Record<string, SchemaRef>} */
  const properties = {};
  for (const [name, property] of Object.entries(resolved.properties ?? {})) {
    properties[name] = resolveSchema(property);
  }
  return { properties, required: new Set(resolved.required ?? []) };
}

/** @param {any} op @returns {SchemaRef|null} */
function responseSchema(op) {
  for (const status of ["200", "201"]) {
    const schema = op.responses?.[status]?.content?.["application/json"]?.schema;
    if (schema) return resolveSchema(schema);
  }
  return null;
}

/** @param {any} document @param {any} schema */
function dereference(document, schema) {
  if (!schema.$ref) return schema;
  return document.components.schemas[schema.$ref.split("/").pop()];
}

/**
 * One schema node → a {@link SchemaRef}.
 *
 * The `type` field is where .NET's document needs care: an integer property is emitted as
 * `type: ["integer", "string"]` with an integer `format`, even though the wire only ever carries a
 * number. Taking that union literally would give every `total` in every SDK an `object` type.
 * `openapi-typescript` already resolves the same union to `number` for the console, so this
 * resolves it the same way — one answer about this document, not one per generator.
 *
 * @param {any} schema
 * @returns {SchemaRef}
 */
export function resolveSchema(schema) {
  if (schema?.$ref) {
    return { componentName: schema.$ref.split("/").pop(), type: "object", itemType: null, format: null };
  }

  const raw = schema?.type;
  const types = typeof raw === "string" ? [raw] : Array.isArray(raw) ? raw : ["object"];
  const type = types.find((t) => t !== "null" && t !== "string") ?? types[0];

  return {
    componentName: null,
    type,
    format: schema?.format ?? null,
    itemType: type === "array" && schema?.items ? resolveSchema(schema.items) : null,
  };
}
